### ISSUE TITLE
Sometimes white screen when using Trusted Web Activity

### ISSUE THREAD

**Issue by emilvandenberg** [openid/AppAuth-Android#569]
I tried to implement Trusted Web Activity, as discussed in #292; however, sometimes the browser will just open with a white screen.

I added the details about the issue as comment to #292, but I think it doesn't get much attention, because it's an old issue. So I now try this way. Please excuse if this is not the right way to raise an issue.

**Comment by GSpanakos** [comment]
I have been trying to solve this for the past few days. From what I gathered the culprit is this constant
`customTabsIntent.Intent.PutExtra(TrustedWebUtils.ExtraLaunchAsTrustedWebActivity, true);`

that enables the url bar to hide. I later pass this on PerformAuthorizationRequest(), to launch my own IDP (IdentityServer 4)
`_authService.PerformAuthorizationRequest(authRequest, postAuthorizationIntent, customTabsIntent);`

and as you've noticed the redirect never happens (white screen) if the user is already logged-in from the device's browser. Leave out the .PutExtra part and everything works as a charm but the url bar is not hidden 😒. My guess is that the problem lies with the signin-redirect.js that Identity Server 4 utilizes. 

Another workaround to what you suggested in #292 would probably be to customize IDS4 with an intermediate page between successful login and the redirect, just like the prompt you get during logout. (Logout works as expected in all cases)

**Comment by emilvandenberg** [comment]
Thanks for the suggestions, I'll have a look.

**Comment by GSpanakos** [comment]
One more, you can set the prompt to "login" on your AuthorizationRequest object to always ask for credentials when your IDP is launched. That solved it for me, it just means you need to re-login (from your device) if by any chance your user is still logged-in from your IDS4 site.

[Prompt](https://identityserver4.readthedocs.io/en/latest/endpoints/authorize.html) parameter from IDS4 docs.

**Comment by emilvandenberg** [comment]
Thanks, that sounds like a good workaround indeed. Still, it means that you have to login again, whereas without TWA it would login by itself, because the user is still logged in on the IDS4 site.

**Comment by santipg382s** [comment]
> Thanks, that sounds like a good workaround indeed. Still, it means that you have to login again, whereas without TWA it would login by itself, because the user is still logged in on the IDS4 site.

**Comment by GSpanakos** [comment]
Indeed, but if you configure your mobile client for offline access with refresh tokens, it is a minor incovinience. Your user would be re-authenticated in the background and depending on the lifetime of the token it could be weeks before he's forced to login again.

**Comment by kawomm** [comment]
Hi, there,

if I understood you right, you did some tries with TWA. Three questions:

1. You faced the same challenges with the need of user interactions? So a weblogin session based approach would alwas lead to a user interaction? Or you use REFRESHTOKENS, as described by @GSpanakos. To my personal understanding it's much worse to store long-lasting REFRESHTOKENS on a public device in comparison to a non-gesture-based fetch of an URL by an activity.
2. What did you change in the app-auth-library to enable the full-screen TWAs? From security perspective it very good to have a URL on the screen, but from marketing perspective people just want to see the login screen an a more or less native manner.
3. Is there a pull-request? 

Thanks in advance.
